### ISSUE TITLE
Rule proposal: report dangling `.then` calls in async functions

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#9624]
**Please describe what the rule should do:**

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[x] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
// invalid:
async function invalid() {
  foo().then(bar);

  foo().then(result => {
    bar(result);
  });

  foo().then(result => {
    bar(result);
  }).catch(doSomethingElse);
}
```

```js
// valid:
async function valid() {
  await foo().then(bar);

  await foo().then(result => {
    bar(result);
  });

  return foo().then(result => {
    bar(result);
  }).catch(doSomethingElse); // (unsure about this case -- maybe `catch` could be opt-out)
}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This rule would flag a common error that developers make when they start working with promises and async functions. The `invalid` examples are typically bugs, because:

* If the `foo()` promise rejects, the error cannot be caught anywhere and might silently be ignored depending on the environment.
* The promise returned by `foo().then(...)` might still be pending when the surrounding function returns.
* In most cases, developers intend for errors to be handled somewhere and not silently ignored.
* If the developer actually did intend to swallow errors, they should make that intention explicit by using `.catch` or a comment, so that the code doesn't look like it has a bug.

**Comment by ljharb** [comment]
I don't think this is a good rule; using "dangling .then" calls is an entirely valid way to interact with promises. Additionally, this forces my `.then` call to block the rest of my async function (which in the "bad" example was explicitly why I'd have chosen not to use `await`), *and* propagates exceptions up to the caller (which i didn't want).

The latter can be addressed by adding `.catch()`, but the former can't be addressed.

At the very least, I don't think that this belongs in core.

**Comment by not-an-aardvark** [comment]
In the case where you want to do something without blocking your async function and handle errors later, you could do this (which wouldn't be reported by the rule):

```js
async function x() {
  const result = foo().then(bar);
  // ...do something with `result` later that would cause errors to be caught
}
```

I won't argue about whether intentionally swallowing errors is a good way to use promises. However, *unintentionally* swallowing errors is a frequent source of bugs if a developer doesn't understand how to use promises. Additionally, it's much more common to want to propagate errors than to want to ignore them (the latter would only be the case for functions with side-effects where the result doesn't matter).

If I see code like `foo().then(bar);` in an async function without `await` or `return`, my first impression is that the code looks like it has a bug. If the intent of the code is actually to swallow exceptions, and there's no comment indicating this, then in my opinion the code has a readability issue and would benefit from adding an explicit `catch` clause or `eslint-disable` directive anyway.

**Comment by ljharb** [comment]
Right, but what if no `result` is desired? Forcing me to create a binding just to achieve the semantics that perfectly valid Promise usage already achieves, which would then violate `no-unused-vars`, doesn't seem like a productive rule.

Is the sole motivation for this rule your personal and subjective viewpoint that *your* first impression is that it's a bug? That certainly isn't mine.

**Comment by not-an-aardvark** [comment]
> Is the sole motivation for this rule your personal and subjective viewpoint that *your* first impression is that it's a bug? That certainly isn't mine.

The main motivation for this rule is my suspicion that in most cases, code like this is *actually* a bug. Based on this suspicion, I think it's worth flagging code like this to the user.

Separately, I'm arguing that the workarounds for this lint error would be worthwhile for readability purposes anyway (based on my personal views about readability, which are anecdotally informed by bugs like this that I've seen in the past). So I'm unconvinced that the use cases you're describing are a good argument against adding this rule.

**Comment by mysticatea** [comment]
Hmm... I'm not sure if this rule is valuable because I think it's rare cases that people use a mix of `.then` and `await`. I probably support if the rule disallows all `.then` in async functions, but...

About C#, Visual Studio reports function calls that their return type is `Task`/`Task<T>` (E.g. async methods) if the call is not in `await` expressions. The feature has caught async function calls which are without `await`. It's very useful personally.
On the other hand, ESLint does not know the return type of functions. Yes, `.then()` probably returns a `Promise`. But I suspect the value of reporting the restricted case.

**Comment by not-an-aardvark** [comment]
Closing because there doesn't seem to be much interest in this.
